src/analog.rs

```diff
@@ -0,0 +1,50 @@
+//! Analog-digital conversion traits
+
+use nb;
+
+/// This trait denotes an object, e.g. a GPIO pin, that is configured for use as an ADC channel.
+/// Objects of this type are passed to ADC-mode-specific methods.
+pub trait AdcChannel {
+    /// Channel ID type
+    ///
+    /// e.g. u8 for ADC channels 1-n, or a (partid, pinid) tuple for something more complex.
+    type ID;
+
+    /// ID of the channel associated with this pin
+    const CHANNEL: Self::ID;
+}
+
+/// ADCs that sample on single channels per request, and do so at the time of the request.
+pub trait SingleMode<Word, Pin: AdcChannel> {
```

Comment:
Minor bikeshedding: `analog::SingleMode` doesn't seem like the a great name here -- and it doesn't really line up with `analog::Continuous`.

Some off-the-cuff ideas:

- `analog::OnDemand`
- `analog::Once`

Reply:
I like both suggestions. Another one: `OneShot`, which I've seen used in this context.

Reply:
Yeah, `OneShot` sampling is the typical term here. I think it's also important to recognise what is being described here: Is this the acquisition or the trigger mode? At the moment it sounds more like a mix of the two.

- `Continuous`: Continuous sampling with single value capture (Immediate trigger)
- `Once`: Single sampling with single value capture (Immediate trigger)

`Continuous` is only mildly useful as it only saves the reinitialisation cost for the ADC but running an ADC continuously to just capture a value every now and then is rather energy intensive.

`Continuous` mode is far more interesting if you can do automatic sampling of many values at a specific rate into a fixed buffer (e.g. audio signal), rolling values into a circular buffer or automatically trigger a recording of a a single or burst of samples at some event (edge/level triggering on the signal).

Reply:
I think it's fair to say that I'm trying to describe the trigger mode here since it's closest to how the hardware configuration is realized. I do agree that Continuous gets more interesting when you use it in the ways you describe. I don't see a clear path there, and I feel like there's kind of an implied DMA activity too, traits for which are not yet very well-defined.